### DISCUSSION TITLE
How can I troubleshoot start-up time?

### DISCUSSION THREAD

**Discussion by chscott** [PowerShell/PowerShell#24206] [Q&A]
I'm experiencing a periodic issue in which PowerShell will not start in a timely manner. By that, I mean the following:

1. Click the PS 7 icon to initiate start up
2. Verify the console window loads in the UI
3. Verify that `PowerShell 7.4.5` prints in the console window
4. Wait as the cursor blinks on the next line for several minutes, until I see this:

![image](https://github.com/user-attachments/assets/cb04f6ab-201e-4610-bb51-81c58ccb28b2)

I don't know this for certain, but the issue seems to occur after Windows Update patches the system. In all cases, I am unable to reproduce the issue after a reboot of the system. I've tried starting PS with the `-NoProfile` option, but it has no effect.

I'm fishing for ideas on how I can troubleshoot the problem.

**Comment by trackd** [comment]
You could test [Profiler](https://github.com/nohwnd/Profiler)  
But its probably not gonna show you alot if `-NoProfile` doesnt have any affect.

If its not a profile thing my next guess would be your antivirus software.
Try and see if the delay is still there if you temporarily disable AV.

**Reply by ghost** [reply]
Ah! I forgot this.

**Comment by SamErde** [comment]
> I'm experiencing a periodic issue in which PowerShell will not start in a timely manner. By that, I mean the following:
> 
> 1. Click the PS 7 icon to initiate start up
> 
> 2. Verify the console window loads in the UI
> 
> 3. Verify that `PowerShell 7.4.5` prints in the console window
> 
> 4. Wait as the cursor blinks on the next line for several minutes, until I see this:
> 
> ![image](https://github.com/user-attachments/assets/cb04f6ab-201e-4610-bb51-81c58ccb28b2)
[...]
> 
> I'm fishing for ideas on how I can troubleshoot the problem.

What do you have in your PowerShell profile? 

Did you by any chance add the line `Import-Module Microsoft.Graph` or `Import-Module Az`? 😀

**Comment by rhubarb-geek-nz** [comment]
Can you run pwsh.exe directly from cmd.exe to test start-up and take Windows icons and Windows Terminal out of the equation?

**Comment by chscott** [comment]
I stopped having the issue some time ago, for reasons unknown. If there is a general checklist of things to do when problems like this occur, that would be useful information to add to this discussion.

**Comment by AbishekPonmudi** [comment]
I've noticed similar startup delays with PowerShell, particularly after Windows updates. One thing that helped me was checking my PowerShell profile for unnecessary module imports. In my case, removing an old Azure module improved startup time. Also, running pwsh.exe directly from cmd.exe gave me a clearer view of the issue without UI-related factors. Has anyone else found a consistent fix?

**Comment by powercode** [comment]
There is a script here: https://github.com/PowerShell/PowerShell/blob/master/tools/performance/Invoke-PerfviewPS.ps1


It requires PerfView.exe, which can be installed by `winget install perfview`.

It will record an ETW trace that can be used to determine the cause of the long startup time.

Ping me if you need help analyzing the file. Windows Performance Analyzer is designed to do that.

```powershell
 .\Invoke-PerfviewPS.ps1 -ScriptBlock {echo hello}
wpa .\PerfViewData.etl
```
in WPA, apply the profile `PowerShell.wpaProfile` from the same directory as the script.

This will give you a view similar to this:

![image](https://github.com/user-attachments/assets/dd90cfff-8858-4044-8a72-1df8390c291e)
